### INTER QUARTILE RANGE (IQR) - OUTLIER DETECTION TECHNIQUE

In [1]:
import pandas as pd
import numpy as np

In [2]:
dfcars=pd.read_csv(r"C:\Users\admin\Desktop\DATA SCIENCE\Machine Learning\machine learning dataset\One Hot Encoding\cars.csv")

In [3]:
dfcars

brand  km_driven    fuel                 owner  selling_price
0      Maruti     145500  Diesel           First Owner         450000
1       Skoda     120000  Diesel          Second Owner         370000
2       Honda     140000  Petrol           Third Owner         158000
3     Hyundai     127000  Diesel           First Owner         225000
4      Maruti     120000  Petrol           First Owner         130000
...       ...        ...     ...                   ...            ...
8123  Hyundai     110000  Petrol           First Owner         320000
8124  Hyundai     119000  Diesel  Fourth & Above Owner         135000
8125   Maruti     120000  Diesel           First Owner         382000
8126     Tata      25000  Diesel           First Owner         290000
8127     Tata      25000  Diesel           First Owner         290000

[8128 rows x 5 columns]

#### we are finding out the outliers in this datset there are many techniqus 
#### we are going to use IQR (Inter Quartile Range)

In [4]:
dfcars.columns

Index(['brand', 'km_driven', 'fuel', 'owner', 'selling_price'], dtype='object')

#### Here you can see we have five columns so inly from Numercial columns you can find outliers,
#### So here we will find out Outlier from Km_driven , selling Price 

In [5]:
dfcars

brand  km_driven    fuel                 owner  selling_price
0      Maruti     145500  Diesel           First Owner         450000
1       Skoda     120000  Diesel          Second Owner         370000
2       Honda     140000  Petrol           Third Owner         158000
3     Hyundai     127000  Diesel           First Owner         225000
4      Maruti     120000  Petrol           First Owner         130000
...       ...        ...     ...                   ...            ...
8123  Hyundai     110000  Petrol           First Owner         320000
8124  Hyundai     119000  Diesel  Fourth & Above Owner         135000
8125   Maruti     120000  Diesel           First Owner         382000
8126     Tata      25000  Diesel           First Owner         290000
8127     Tata      25000  Diesel           First Owner         290000

[8128 rows x 5 columns]

In [6]:
dfcars["km_driven"].describe()

count    8.128000e+03
mean     6.981951e+04
std      5.655055e+04
min      1.000000e+00
25%      3.500000e+04
50%      6.000000e+04
75%      9.800000e+04
max      2.360457e+06
Name: km_driven, dtype: float64

see guys here 25% means q1
             
             
            75% means q3

In [7]:
q1=dfcars["km_driven"].quantile(0.25)

In [8]:
q1

35000.0

In [9]:
q3=dfcars["km_driven"].quantile(0.75)

In [10]:
q3

98000.0

same values are coming that of 25% and 75% of describe
dont get panic in describe 25% it 3.500000e+04 its mantisa 
its the same value

In [11]:
9.8000*10000

98000.0

as you can now have a look at it

In [12]:
q1

35000.0

In [13]:
q3

98000.0

In [14]:
iqr=q3-q1

In [15]:
iqr

63000.0

In [16]:
dfcars

brand  km_driven    fuel                 owner  selling_price
0      Maruti     145500  Diesel           First Owner         450000
1       Skoda     120000  Diesel          Second Owner         370000
2       Honda     140000  Petrol           Third Owner         158000
3     Hyundai     127000  Diesel           First Owner         225000
4      Maruti     120000  Petrol           First Owner         130000
...       ...        ...     ...                   ...            ...
8123  Hyundai     110000  Petrol           First Owner         320000
8124  Hyundai     119000  Diesel  Fourth & Above Owner         135000
8125   Maruti     120000  Diesel           First Owner         382000
8126     Tata      25000  Diesel           First Owner         290000
8127     Tata      25000  Diesel           First Owner         290000

[8128 rows x 5 columns]

In [17]:
dfcars["km_driven"]

0       145500
1       120000
2       140000
3       127000
4       120000
         ...  
8123    110000
8124    119000
8125    120000
8126     25000
8127     25000
Name: km_driven, Length: 8128, dtype: int64

In [18]:
lower=q1 - (1.5 * iqr)

In [19]:
lower

-59500.0

In [20]:
upper = q3 + (1.5 * iqr)

In [21]:
upper

192500.0

Now you can see our lower value is -59500 so below -59000 all the data points are considered as oulier sames goes for upper

## TRIMMING

In [23]:
dfcars[ (dfcars["km_driven"] < lower) | (dfcars["km_driven"] > upper) ]

brand  km_driven    fuel         owner  selling_price
272      Maruti     193000  Diesel   First Owner         320000
287    Mahindra     207890  Diesel  Second Owner         210000
394      Toyota     240000  Diesel   Third Owner         250000
396     Hyundai     214000  Petrol  Second Owner         340000
397     Hyundai     214000  Diesel  Second Owner         340000
...         ...        ...     ...           ...            ...
8031  Chevrolet     195000  Diesel   First Owner         275000
8033    Hyundai     200000  Diesel  Second Owner         450000
8071     Toyota     200000  Diesel  Second Owner         425000
8077     Toyota     250000  Diesel   First Owner         550000
8095       Ford     194000  Diesel   First Owner         260000

[170 rows x 5 columns]

Now you can see their are 170 Rows which are considered as oulier 
so Either you can remove it or make them to your lower limit and upper limit which is considered as capping

In [24]:
dfcar1=dfcars[ (dfcars["km_driven"] < lower) | (dfcars["km_driven"] > upper) ]

Now we have put all these thing in a object

lets remove it first

In [40]:
dfcar2=dfcars.copy(deep=False)

we have made a copy of our main dataset so that our main dataset should not get damaged

In [41]:
dfcar2=pd.DataFrame(dfcar2)

In [42]:
dfcar2

brand  km_driven    fuel                 owner  selling_price
0      Maruti     145500  Diesel           First Owner         450000
1       Skoda     120000  Diesel          Second Owner         370000
2       Honda     140000  Petrol           Third Owner         158000
3     Hyundai     127000  Diesel           First Owner         225000
4      Maruti     120000  Petrol           First Owner         130000
...       ...        ...     ...                   ...            ...
8123  Hyundai     110000  Petrol           First Owner         320000
8124  Hyundai     119000  Diesel  Fourth & Above Owner         135000
8125   Maruti     120000  Diesel           First Owner         382000
8126     Tata      25000  Diesel           First Owner         290000
8127     Tata      25000  Diesel           First Owner         290000

[8128 rows x 5 columns]

lets remove all the outliers now

we will remove from dfcar2 bcoz we also wanted to capping right 
Lets do it

In [48]:
dfcar2= dfcar2[(dfcar2["km_driven"] >= lower) & (dfcar2["km_driven"] <= upper)]

In [49]:
dfcar2

brand  km_driven    fuel                 owner  selling_price
0      Maruti     145500  Diesel           First Owner         450000
1       Skoda     120000  Diesel          Second Owner         370000
2       Honda     140000  Petrol           Third Owner         158000
3     Hyundai     127000  Diesel           First Owner         225000
4      Maruti     120000  Petrol           First Owner         130000
...       ...        ...     ...                   ...            ...
8123  Hyundai     110000  Petrol           First Owner         320000
8124  Hyundai     119000  Diesel  Fourth & Above Owner         135000
8125   Maruti     120000  Diesel           First Owner         382000
8126     Tata      25000  Diesel           First Owner         290000
8127     Tata      25000  Diesel           First Owner         290000

[7958 rows x 5 columns]

### Insights:-
So you can see now our rows were 8128 now we have 7958 rows it has
removed all the outliers but see guys what the worrysome thing here
think like you have less no of the datapoints and more outliers are their what you are going to do then 

we have one method for that which is Capping

In [50]:
dfcars

brand  km_driven    fuel                 owner  selling_price
0      Maruti     145500  Diesel           First Owner         450000
1       Skoda     120000  Diesel          Second Owner         370000
2       Honda     140000  Petrol           Third Owner         158000
3     Hyundai     127000  Diesel           First Owner         225000
4      Maruti     120000  Petrol           First Owner         130000
...       ...        ...     ...                   ...            ...
8123  Hyundai     110000  Petrol           First Owner         320000
8124  Hyundai     119000  Diesel  Fourth & Above Owner         135000
8125   Maruti     120000  Diesel           First Owner         382000
8126     Tata      25000  Diesel           First Owner         290000
8127     Tata      25000  Diesel           First Owner         290000

[8128 rows x 5 columns]

### we will see capping in the next dataset